gcc/rust/resolve/rust-name-resolver-2.0.h

```diff
@@ -0,0 +1,181 @@
+new "Scope" to a stack of these scopes, and "pops" it when exiting said lexical
+scope. However, because we are splitting the name resolution into two passes, we
+would like to avoid re-doing a bunch of work we've already done - which is why
+this data structure needs to allow "re-entrancy", or to at least not keep as
+much state as the existing one, and allow for viewing the same module multiple
+times without throwing a fit.
+
+  We will be implementing a "forever stack" of scopes, which allows the user the
+pushing of new scopes onto the stack, but only simulates the popping of a scope:
+When pushing new scopes, more space is allocated on our stack, and we keep
+track of this scope as being the current one - however, when popping this scope,
+we do not actually delete the memory associated with it: we simply mark the
+previous scope (parent) as the current one.
+
+In the example below, each number indicates the "state" of our resolution map,
+and the carret is used to point to the current lexical scope.
+
+```rust
+		// []
+		//
+fn main() {     // [ `main` scope: {} ]
+		//         ^
+  let a = 15;   // [ `main` scope: { Decl(a) } ]
+		//         ^
+  {  _PUSH_     // [ `main` scope: { Decl(a) }, anonymous scope: {} ]
+		//                                        ^
+    let a = 16; // [ `main` scope: { Decl(a) }, anonymous scope: { Decl(a) } ]
+		//                                        ^
+    f(a);       // [ `main` scope: { Decl(a) }, anonymous scope: { Decl(a) } ]
+		//                                        ^
+  }   _POP_     // [ `main` scope: { Decl(a) }, anonymous scope: { Decl(a) } ]
+		//         ^
+  f(a);         // [ `main` scope: { Decl(a) }, anonymous scope: { Decl(a) } ]
+		//         ^
+}
+```
+
+This allows us to revisit scopes previously visited in later phases of the name
+resolution, and add more information if necessary.
+
+  2. `Late`
+
+  `Late` name resolution possesses some unique challenges since Rust's name
+resolution rules are extremely complex - variable shadowing, variable capture in
+closures (but not inner functions!)... You can have a look at a fucked up
+example here:
+
+https://rustc-dev-guide.rust-lang.org/name-resolution.html#scopes-and-ribs
+
+This requires us to think about what exactly to put in our `Scope`s and what to
+do with our `Rib`s - and how it affects our data structures. For example, in the
+above example, `rustc` demonstrates how multiple `Rib`s can be created inside of
+a single lexical scope for variables, as the Rust programming language allows
+shadowing.
+
+    TODO: Mention macro hygiene and that it is the same
+    TODO: How does this affect our data structures?
+    TODO: Last challenge - reuse the same APIs to allow the typechecker to not
+change?
+    TODO: Mention that ForeverStack is templated to make sure that behavior is
+correct
+*/
+
+// Now our resolver, which keeps track of all the `ForeverStack`s we could want
+class Resolver
```

Comment:
Having a class called `Resolver` that does not inherit from `DefaultResolver` is a bit confusing.

Reply:
That's a good point. I'll rename it to `ResolverContext`, which is different from what we had but makes more sense IMO